### DISCUSSION TITLE
cpptools 1.8.4 scans whole disk from root on each build cycle. Why ?

### DISCUSSION THREAD

**Discussion by ieb** [microsoft/vscode-cpptools#8800] [Q&A]
I am running PlatformIO on OSX with about 4 projects, all small perhaps 10 files in each. When I save a 1 byte change to a cpp or h file cpptools kicks of for about 15m running at 80% - 100% CPU with the Fans at full RPM. A build of one of these projects from scratch takes about 120s. I know cpptools needs to do more than a build but the activity and fans are continuous so I started to look at what it does.

If I monitor the files cpptools is opening (watching open files in Activity Monitor), it does a full scan of my entire hard disk on every build. It looks like it is searching for c or header files and since I have a large Brew Cellar tree in /usr/local it finds a lot of files, none of which have anything to do with VSCode or the project I am working on.

Is there a way of stopping cpptools from scanning my entire disk and limiting itself to the files it should be considering ?

I did notice that PlatformIO appends an empty string, ie "" to the include path in c_cpp_properties.json which could be related ?

I've searched high and low for a solution in discussions, git issues and generally but nothing seems to make any difference.

Update:
Patching the template used in PlatformIO to generate the c_cpp_properties.json removing the trailing "" on include makes no difference. cpptools continues to scan the entire disk.

Deleting the sqllite databases in ~/Library/Application Support/Code/User/workspaceStorage/<UUID>/ms-vscode.cpptools/** doesn't change the behaviour.

PlatformIO is uptodate, so I suspect this behaviour might be by design for cpptools, but honestly, its not right.

**Comment by ieb** [comment]
Setting C_Cpp.intelliSenseEngine to "Disabled" in the workspace configuration fixes the problem, but also disables intellisense. TBH, being able to type code in near real time is more important to me than being told the signatures I already know, (or have nothing to do with the processor architecture being worked on).

**Reply by sean-mcmanus** [reply]
You should be able to fix this by adding to setting "C_Cpp.files.exclude" (e.g. `"/usr/local/folder": true` or `"**/folder": true`. Let us know if that isn't working.

**Reply by ieb** [reply]
I tried adding the following to the  code-workspace file. I think it has fixed it. Thank you.

cpptools did scan the whole disk on startup, but seems to now not scan the whole disk on every change to a cpp or h file.




```
{ 
...
  "settings": {
...
    "C_Cpp.files.exclude" : {
                 "/usr": true,
                 "/Applications": true,
                 "/User Guides And Information": true,
                 "cores": true,
                 "/Library": true,
                 "/dev": true,
                 "/net" :true,
                 "/bin" : true,
                 "/home": true,
                 "/private": true,
                 "/var": true,
                 "/System": true,
                 "/sbin": true
          }
     }
```

**Reply by Colengms** [reply]
Hi @ieb .  The only directories the C/C++ Extension should be scanning are within the workspace folder(s) you have open, directories specified as 'include' paths in `c_cpp_properties.json` (or other source of configurations), and any include paths returned from the compiler when queried.  Is it possible that one of these directories contains a symbolic link that points up to the root of your HD?  If so, removing that would seem to be the fix.  In a simple example on a stock system, I don't believe we are seeing the extension try to scan entire HDs.

**Reply by sean-mcmanus** [reply]
If your paths you're trying to exclude involve symlinks you could be hitting https://github.com/microsoft/vscode-cpptools/issues/2375 .

**Reply by ieb** [reply]
The c_cpp_properties.json file is generated by PlatformIO, and only contains paths for the workspace and platform (eg ESP32).  It does include a empty quoted string as the last entry in the list `""`. Removing that doesn't change the behaviour.

The locations in include do have symbolic links but nothing outside of the same tree, no recursions and no link back to /.  I visually inspected the links but it is always possible I missed following  a second or third level link that jumped back to / 

Monitoring lsof or the activity monitor list files and ports without the fix to C_Cpp.files.exclude shows the cpptools process scanning the whole disk on startup and on every save operation. All of the paths I excluded appear at some point.
 
With the fix, I see evidence of a full disk scan on startup, but not on each save.
With the fix, after a few hours of work VSCode + Intellisense seems to be operating perfectly normally, detecting problems and showing the correct versions of functions on hover over.
